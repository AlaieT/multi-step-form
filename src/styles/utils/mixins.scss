@mixin flex($param: null, $type: null) {
  display: flex;

  @if  $param == "center&center" {
    justify-content: center;
    align-items: center;
  }

  @if  $param == "center&end" {
    justify-content: center;
    align-items: flex-end;
  }

  @if  $param == "end&center" {
    justify-content: flex-end;
    align-items: center;
  }

  @if  $param == "space-between&center" {
    justify-content: space-between;
    align-items: center;
  }

  @if  $type == "col" {
    flex-direction: column;
  }
}
