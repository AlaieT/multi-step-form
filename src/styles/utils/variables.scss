$background-white: rgb(255 255 255);
$background-global: rgb(235 243 254);
$background-select-step: rgb(191 226 253);
$primary-text-color: rgb(2 41 90);
$secondary-text-color: rgb(150 153 171);
$plan-border-color: rgb(85 75 209);
$summary-hover-button: rgb(133 127 250);
$select-input: rgb(71 61 255);
$select-button: rgb(0 62 138);
$select-plan: rgb(246 247 254);
$error: hsl(354deg 84% 57%);
$shadow-color: rgb(214 217 230);
$content-box-shadow: 0 25px 40px -20px rgba(0 0 0 / 10%);
