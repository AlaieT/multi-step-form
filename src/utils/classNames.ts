const classNames = (...classes: (string | undefined)[]): string | undefined => {
  return classes.filter((item) => !!item).join(" ");
};

export default classNames;
