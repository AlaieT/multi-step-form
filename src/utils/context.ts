import { createContext } from "react";

import type { FormContextType } from "../types";

const FormContext = createContext<FormContextType>({
  form: {
    planMode: "yearly",
    priceMode: "yr"
  },
  setForm: () => undefined
});

export { FormContext };
