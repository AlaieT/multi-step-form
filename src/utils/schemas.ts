import { object, string, boolean, number } from "yup";

const multiStepFromSchema = object({
  name: string().required("Name is a required field!"),
  email: string()
    .required("Email is a required field!")
    .email("Must be a valid email!"),
  phone: string().required("Phone is a required field!"),
  plan: string().oneOf(["arcade", "advanced", "pro"]).defined(),
  planMode: boolean().defined(),
  onlineService: boolean().defined(),
  largeStorage: boolean().defined(),
  customizableProfile: boolean().defined(),
  totalPrice: number().defined(),
});

export { multiStepFromSchema };
