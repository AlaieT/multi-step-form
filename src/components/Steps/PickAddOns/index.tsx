import React, { useContext } from "react";

import Button from "../../Button";
import Wrap from "../Wrap";

import type { PickAddOnsProps } from "../../../types";

import styles from "../../../styles/components/steps/pickAddOns.module.scss";
import { FormContext } from "../../../utils/context";

const PickAddOns = ({
  addOns: { onlineService, largeStorage, customizableProfile },
  register,
  getValues,
  changeStep
}: PickAddOnsProps) => {
  const { form } = useContext(FormContext);

  return (
    <Wrap>
      <div id={styles.content}>
        <div id={styles.header}>
          <h1>Pick add-ons</h1>
          <h2>Add-ons help enhance your gaming experience.</h2>
        </div>
        <div className={styles.option}>
          <input
            id="onlineService"
            type="checkbox"
            defaultChecked={getValues("onlineService")}
            {...register("onlineService")}
          />
          <span />
          <div className={styles.title}>
            <label htmlFor="onlineService">Online service</label>
            <label htmlFor="onlineService">Access to multiplayer games</label>
          </div>
          +$
          {`${onlineService[form.planMode]}/${form.priceMode}`}
        </div>
        <div className={styles.option}>
          <input
            id="largeStorage"
            type="checkbox"
            defaultChecked={getValues("largeStorage")}
            {...register("largeStorage")}
          />
          <span />
          <div className={styles.title}>
            <label htmlFor="largeStorage">Larger storage</label>
            <label htmlFor="largeStorage">Extra 1TB of cloud save</label>
          </div>
          +$
          {`${largeStorage[form.planMode]}/${form.priceMode}`}
        </div>
        <div className={styles.option}>
          <input
            id="customizableProfile"
            type="checkbox"
            defaultChecked={getValues("customizableProfile")}
            {...register("customizableProfile")}
          />
          <span />
          <div className={styles.title}>
            <label htmlFor="customizableProfile">Customizable profile</label>
            <label htmlFor="customizableProfile">
              Custom theme on your profile
            </label>
          </div>
          +$
          {`${customizableProfile[form.planMode]}/${form.priceMode}`}
        </div>
      </div>
      <div id={styles.controls}>
        <Button id={styles.go_back} type="button" onClick={() => changeStep(1)}>
          Go Back
        </Button>
        <Button
          id={styles.next_step}
          type="button"
          onClick={() => changeStep(3)}
        >
          Next Step
        </Button>
      </div>
    </Wrap>
  );
};

export default PickAddOns;
