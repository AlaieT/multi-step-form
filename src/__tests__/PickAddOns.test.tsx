import React from "react";
import {
  render,
  cleanup,
  screen,
  fireEvent,
  renderHook
} from "@testing-library/react";
import { useForm } from "react-hook-form";

import PickAddOns from "../components/Steps/PickAddOns";
import { FormContext } from "../utils/context";

afterEach(cleanup);

describe("PickAddOns", () => {
  const mockChangeStep = jest.fn();
  const props = global.getStepProps();

  describe("correct render", () => {
    it("should match snapshot", () => {
      const { result } = renderHook(() => useForm(global.getUseFormProps()));

      expect(
        render(
          <FormContext.Provider
            value={{
              form: {
                planMode: "yearly",
                priceMode: "yr"
              },
              setForm: () => undefined
            }}
          >
            <PickAddOns
              addOns={props.addOns}
              register={result.current.register}
              getValues={result.current.getValues}
              changeStep={mockChangeStep}
            />
          </FormContext.Provider>
        ).asFragment()
      ).toMatchSnapshot();
    });
  });

  describe("correct functionality", () => {
    it("should correct select add-ons", () => {
      const { result } = renderHook(() => useForm(global.getUseFormProps()));

      render(
        <FormContext.Provider
          value={{
            form: {
              planMode: "yearly",
              priceMode: "yr"
            },
            setForm: () => undefined
          }}
        >
          <PickAddOns
            addOns={props.addOns}
            register={result.current.register}
            getValues={result.current.getValues}
            changeStep={mockChangeStep}
          />
        </FormContext.Provider>
      );

      expect(screen.getByLabelText(/online service/i)).not.toBeChecked();
      expect(screen.getByLabelText(/larger storage/i)).not.toBeChecked();
      expect(screen.getByLabelText(/customizable profile/i)).not.toBeChecked();

      fireEvent.click(screen.getByLabelText(/online service/i));
      fireEvent.click(screen.getByLabelText(/larger storage/i));
      fireEvent.click(screen.getByLabelText(/customizable profile/i));

      expect(screen.getByLabelText(/online service/i)).toBeChecked();
      expect(screen.getByLabelText(/larger storage/i)).toBeChecked();
      expect(screen.getByLabelText(/customizable profile/i)).toBeChecked();
    });
  });
});
