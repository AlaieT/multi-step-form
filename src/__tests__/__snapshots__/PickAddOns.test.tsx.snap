// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickAddOns correct render should match snapshot 1`] = `
<DocumentFragment>
  <div
    id="step"
  >
    <div
      id="content"
    >
      <div
        id="header"
      >
        <h1>
          Pick add-ons
        </h1>
        <h2>
          Add-ons help enhance your gaming experience.
        </h2>
      </div>
      <div
        class="option"
      >
        <input
          id="onlineService"
          name="onlineService"
          type="checkbox"
        />
        <span />
        <div
          class="title"
        >
          <label
            for="onlineService"
          >
            Online service
          </label>
          <label
            for="onlineService"
          >
            Access to multiplayer games
          </label>
        </div>
        +$10/yr
      </div>
      <div
        class="option"
      >
        <input
          id="largeStorage"
          name="largeStorage"
          type="checkbox"
        />
        <span />
        <div
          class="title"
        >
          <label
            for="largeStorage"
          >
            Larger storage
          </label>
          <label
            for="largeStorage"
          >
            Extra 1TB of cloud save
          </label>
        </div>
        +$20/yr
      </div>
      <div
        class="option"
      >
        <input
          id="customizableProfile"
          name="customizableProfile"
          type="checkbox"
        />
        <span />
        <div
          class="title"
        >
          <label
            for="customizableProfile"
          >
            Customizable profile
          </label>
          <label
            for="customizableProfile"
          >
            Custom theme on your profile
          </label>
        </div>
        +$30/yr
      </div>
    </div>
    <div
      id="controls"
    >
      <button
        class="button"
        id="go_back"
        type="button"
      >
        Go Back
      </button>
      <button
        class="button"
        id="next_step"
        type="button"
      >
        Next Step
      </button>
    </div>
  </div>
</DocumentFragment>
`;
